pkg/aws/client/updater.go

```diff
@@ -102,54 +103,73 @@ func (u *updater) UpdateSecurityGroup(ctx context.Context, desired, current *Sec
 	return true, nil
 }
 
-func (u *updater) UpdateRouteTable(ctx context.Context, log logr.Logger, desired, current *RouteTable, controlledCidrBlocks ...string) (modified bool, err error) {
-outerDelete:
-	for _, cr := range current.Routes {
-		for _, dr := range desired.Routes {
-			if reflect.DeepEqual(cr, dr) {
-				continue outerDelete
-			}
+// TODO: Consider IPv4 xor IPv6 xor destination prefix lists as destination.
+func (u *updater) UpdateRouteTable(ctx context.Context, log logr.Logger, desired, current *RouteTable) (bool, error) {
+	var (
+		routesToDelete = []*Route{}
+		routesToCreate = []*Route{}
+		modified       = false
+	)
+
+	for _, r := range current.Routes {
+		if r == nil {
+			continue
 		}
-		if cr.GatewayId != nil && *cr.GatewayId == "local" {
-			// ignore local gateway route
-			continue outerDelete
+		if !slices.Contains(desired.Routes, r) {
+			routesToDelete = append(routesToDelete, r)
 		}
-		if cr.DestinationPrefixListId != nil {
-			// ignore VPC endpoint route table associations
-			continue outerDelete
+	}
+	for _, r := range desired.Routes {
+		if r == nil {
+			continue
 		}
-		routeCidrBlock := ptr.Deref(cr.DestinationCidrBlock, "")
-		found := false
-		for _, cidr := range controlledCidrBlocks {
-			if routeCidrBlock == cidr {
-				found = true
-				break
-			}
+		if !slices.Contains(current.Routes, r) {
+			routesToCreate = append(routesToCreate, r)
 		}
-		if !found {
-			// ignore unknown routes
+	}
+
+	for _, route := range routesToDelete {
+		if route.GatewayId != nil && *route.GatewayId == "local" {
+			// ignore local gateway route
 			continue
 		}
-		if err = u.client.DeleteRoute(ctx, current.RouteTableId, cr); err != nil {
-			return
+
+		deletionCandidateDestination, err := route.DestinationId()
+		if err != nil {
+			log.Error(err, "failed to find suitable destination for deletion candidate route route")
+			return false, err
 		}
-		log.Info("Deleted route", "cidr", routeCidrBlock)
-		modified = true
-	}
-outerCreate:
-	for _, dr := range desired.Routes {
-		for _, cr := range current.Routes {
-			if reflect.DeepEqual(cr, dr) {
-				continue outerCreate
+
+		// The current route table may contain routes not created by the infrastructure controller e.g. the aws-custom-route-controller.
+		// These extra routes will be included in the routesToDelete, but they need to be skipped.
+		if !slices.ContainsFunc(desired.Routes, func(r *Route) bool {
+			desiredRouteDestination, err := r.DestinationId()
+			if err != nil {
+				log.Error(err, "failed to find suitable destination for desired route", "route")
```

Comment:
This will result in a line like `msg="failed to find suitable destination for desired route" error="no route destination found"`, which is a little bit redundant for my taste.
Maybe something more like
```golang
log.Error(err, "error deleting route", "route")
```
Which is not really that much better but does not contain the same information twice.
Also, I think there is an argument missing, don't you mean:
```golang
log.Error(err, "failed to find suitable destination for desired route", "route", route)
```